src/python/pants/backend/python/util_rules/pex.py

```diff
@@ -587,37 +612,87 @@ def _validate_metadata(
     if validation:
         return
 
-    message_parts = [
-        f"Invalid lockfile for PEX request `{request.output_filename}`.",
-        "\n",
-    ]
+    def tool_message_parts(
+        requirements: (ToolCustomLockfile | ToolProgrammaticLockfile | ToolDefaultLockfile),
+    ) -> Iterator[str]:
 
-    if InvalidLockfileReason.INVALIDATION_DIGEST_MISMATCH in validation.failure_reasons:
-        message_parts.append(
-            "The requirements set for this PEX request are different to the requirements set when "
-            "the lockfile was generated. To fix this, you will need to regenerate the lockfile. "
-        )
-        message_parts.append(
-            f"(Expected requirements digest: {requirements.lockfile_hex_digest}, "
-            f"actual: {metadata.requirements_invalidation_digest})"
-        )
-        message_parts.append("\n")
-
-    if InvalidLockfileReason.INTERPRETER_CONSTRAINTS_MISMATCH in validation.failure_reasons:
-        message_parts.append(
-            "The lockfile was generated under interpreter constraints "
-            f"({ metadata.valid_for_interpreter_constraints }) that are incompatible with the "
-            f"constraints set in the project ({request.interpreter_constraints }). If you have "
-            "overridden your project's interpreter constraints, you can update them to specify a "
-            "subset of the interpreters specified in the lockfile. If  not, you will need to "
-            "regenerate your lockfile. "
+        tool_name = requirements.options_scope_name
+        uses_source_plugins = requirements.uses_source_plugins
+        uses_project_interpreter_constraints = requirements.uses_project_interpreter_constraints
+
+        yield "You are using "
+
+        if isinstance(requirements, ToolDefaultLockfile):
+            yield "the `<default>` lockfile provided by Pants "
+        elif isinstance(requirements, ToolProgrammaticLockfile):
+            yield "a lockfile that was generated programmatically "
+        elif isinstance(requirements, ToolCustomLockfile):
+            yield f"the lockfile at {requirements.file_path} "
+        else:
+            raise Exception("Check the type hinting on `requirements`")
```

Comment:
type hints on `requirements` say that this line should never be hit, but I'm not sure mypy agrees.

Reply:
Better is `raise AssertionError`. And yeah this is good to keep. Although note that it will fail much earlier on line 619 due to `AttributeError` if it isn't a `_ToolLockfileMixin`. 

Maybe you can assert that it's an instance of `_ToolLockfileMixin`? It is totally valid to not support invalidation for non-tool lockfiles in Pants 2.7.

Reply:
Yeah, it will definitely fail earlier.

I didn't really want to expose `_ToolLockfileMixin`, it's mostly there to ensure consistency of field names through DRY more than an implementation detail, we can change that if you think it helps.

Reply:
I think it's fine to keep it private w/ the underscore, but still import it here. I strongly suspect this code will change in Pants 2.8 when we add support for multiple user lockfiles - so it makes sense for the type to be an implementation detail.

Reply:
For what it's worth, my expectation is that we'll need a separate error message function to take care of user lockfiles